Instructions for cmake installations without sudo privileges

juanjqo opened the pull request:
Hi @mmmarinho, 

This PR adds a subsection for advanced users (forbidden topics) to install cmake packages without sudo privileges.
Please let me know if additional modifications are required. 

Kind regards, 

Juancho

mmmarinho commented on docs/source/cmake/cmake_packages_without_sudo.rst:
```diff
@@ -0,0 +1,95 @@
+.. code-block:: console
+
+    cmake -DCMAKE_INSTALL_PREFIX:PATH=~/opt .. 
+    make -j16
```
@juanjqo Just in case might be better to leave `-j16` out or explain what the numbers mean otherwise a computer with fewer cores might decide to start a space program or catch on fire.

mmmarinho commented on docs/source/cmake/cmake_packages_without_sudo.rst:
```diff
@@ -0,0 +1,95 @@
+    cd cpp
+    mkdir build && cd build
+    cmake -DCMAKE_INSTALL_PREFIX:PATH=~/opt .. 
+    make -j16
```
`-j16` comment

mmmarinho commented on ros2_tutorial_workspace/src/cpp_cmake_example_dqrobotics/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+               ${PROJECT_NAME}.cpp)
+
+target_link_libraries(${PROJECT_NAME}
+                      dqrobotics)
```
Please be sure that CMake will prefer the local one in case the user already has the global one installed. I suppose that is the case.

mmmarinho commented on docs/source/cmake/cmake_packages_without_sudo.rst:
```diff
@@ -0,0 +1,95 @@
+
+.. code-block:: console
+
+    echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/opt/lib" >> ~/.bashrc
```
According to other parts of the tutorial, please add a command that creates a commented line in `~\.bashrc` explaining where that directive came from.

mmmarinho commented on docs/source/cmake/cmake_packages_without_sudo.rst:
```diff
@@ -0,0 +1,95 @@
+.. code-block:: console
+
+    cd ~/
+    mkdir opt && cd opt
```
some of these might be reused so if you'd prefer you can do
`mkdir -p opt`
and so on.

With this option, `mkdir` will not complain if the folder already exists.

mmmarinho reviewed (changes requested):
Thanks, @juanjqo. Just a few comments!

mmmarinho commented:
For dev packages, you can add the include directory to CPATH.
for example for `~/opt/include`
https://stackoverflow.com/questions/558803/how-can-i-add-a-default-include-path-for-gcc-in-linux

Then the includes will also be automatically found

mmmarinho commented:
Besides `LD_LIBRARY_PATH` I also had to set `LIBRARY_PATH` for it to work.

juanjqo commented:
> Besides `LD_LIBRARY_PATH` I also had to set `LIBRARY_PATH` for it to work.

Thanks. I'm going to update the instructions accordingly.

juanjqo commented:
Hi @mmmarinho,

Please let me know if additional modifications are required.

Kind regards, 

Juancho

juanjqo commented:
Hi @mmmarinho, 

I added the requested modifications. 

- The CMake example is using qpOASES instead of DQ Robotics
- Added instructions to specify the local installation in case the users have more than one.

Kind regards, 

Juancho

mmmarinho commented:
Hi, @juanjqo,

GithubActions didn't like your new package. :(
Any suggestions for `colcon` to work?

```
/usr/bin/bash -c source /opt/ros/humble/setup.sh && colcon build --event-handlers=console_cohesion+ --symlink-install
  [0.692s] ERROR:colcon:colcon build: Duplicate package names not supported:
  - test_proxsuite:
    - src/0mg223k93zdr/ROS2_Tutorial/ros2_tutorial_workspace/src/cpp_cmake_example_qpOASES
    - src/0mg223k93zdr/ROS2_Tutorial/ros2_tutorial_workspace/src/cpp_cmake_example_qpoases_lib
```

juanjqo commented:
> Hi, @juanjqo,
> 
> GithubActions didn't like your new package. :( Any suggestions for `colcon` to work?
> 
> ```
> /usr/bin/bash -c source /opt/ros/humble/setup.sh && colcon build --event-handlers=console_cohesion+ --symlink-install
>   [0.692s] ERROR:colcon:colcon build: Duplicate package names not supported:
>   - test_proxsuite:
>     - src/0mg223k93zdr/ROS2_Tutorial/ros2_tutorial_workspace/src/cpp_cmake_example_qpOASES
>     - src/0mg223k93zdr/ROS2_Tutorial/ros2_tutorial_workspace/src/cpp_cmake_example_qpoases_lib
> ```
> 
> PS: I added the actions to run in the `pull_request` trigger so when you create another pull request we will be able to see the issues before the merge.

@mmmarinho the package `cpp_cmake_example_qpOASES` is deprecated. I forgot to remove it. I did not see the compilation problem using readthedocs. Furthermore, I removed the packaged, but the PR was already merged.

mmmarinho commented:
Thanks, @juanjqo.

I removed the package, but please create another pull request. To address the following.
```
  CMake Error at CMakeLists.txt:36 (message):
    Local DQ Robotics not found!
  
  
  -- Configuring incomplete, errors occurred!
  ---
  --- stderr: test_dqrobotics
  CMake Error at CMakeLists.txt:36 (message):
    Local DQ Robotics not found!
```

Even if it works fine in `ReadTheDocs`, the packages are built in GitHub actions `ros2` environments because that's the only way we can be sure that the code we wrote at least compiles. 

One way to solve this problem is to move the code out of `ros2_tutorial_workspace/src` so that `colcon` stops trying to build it.
